File: src/java/non-corba/com/metricstream/omg/naming/NamingServiceFactory.java
package com.metricstream.omg.naming;

/**
 * Factory class for creating and accessing naming service components.
 * This replaces the CORBA ORB initialization and initial references for the naming service.
 */
public class NamingServiceFactory {
    
    private static NamingServiceFactory instance;
    private NamingContext rootContext;
    
    /**
     * Private constructor to enforce singleton pattern.
     */
    private NamingServiceFactory() {
        // Create the root naming context
        rootContext = new NamingContextImpl();
    }
    
    /**
     * Gets the singleton instance of the factory.
     * 
     * @return The factory instance
     */
    public static synchronized NamingServiceFactory getInstance() {
        if (instance == null) {
            instance = new NamingServiceFactory();
        }
        return instance;
    }
    
    /**
     * Gets the root naming context.
     * 
     * @return The root naming context
     */
    public NamingContext getRootContext() {
        return rootContext;
    }
    
    /**
     * Creates a new naming context.
     * 
     * @return A new naming context
     */
    public NamingContext createNamingContext() {
        return new NamingContextImpl();
    }
    
    /**
     * Resets the naming service, clearing all bindings.
     * This is primarily useful for testing.
     */
    public void reset() {
        rootContext = new NamingContextImpl();
    }
}

File: src/main/kotlin/com/metricstream/naming/NamingServiceFactory.kt
package com.metricstream.naming

import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

/**
 * Factory for creating and accessing naming service components.
 * Replaces the CORBA ORB's resolve_initial_references functionality.
 */
class NamingServiceFactory private constructor() {
    private val rootContext = NamingContextImpl()
    
    /**
     * Gets the root naming context.
     * Equivalent to CORBA's resolve_initial_references("NameService").
     * 
     * @return The root naming context
     */
    fun getRootContext(): NamingContext {
        logger.debug { "Returning root naming context" }
        return rootContext
    }
    
    /**
     * Creates a new naming context.
     * 
     * @return A new naming context
     */
    fun createNamingContext(): NamingContext {
        logger.debug { "Creating new naming context" }
        return NamingContextImpl()
    }
    
    /**
     * Binds an object to a name in the root context.
     * 
     * @param name The name to bind
     * @param obj The object to bind
     */
    fun bind(name: Name, obj: Any) {
        rootContext.bind(name, obj)
    }
    
    /**
     * Resolves a name in the root context.
     * 
     * @param name The name to resolve
     * @return The object bound to the name
     */
    fun resolve(name: Name): Any {
        return rootContext.resolve(name)
    }
    
    /**
     * Resolves a name in the root context using a string representation.
     * Format: "id1.kind1/id2.kind2/id3.kind3"
     * 
     * @param nameStr The string representation of the name
     * @return The object bound to the name
     */
    fun resolveString(nameStr: String): Any {
        val name = Name.fromString(nameStr)
        return resolve(name)
    }
    
    companion object {
        private val INSTANCE = NamingServiceFactory()
        
        /**
         * Gets the singleton instance of the factory.
         * 
         * @return The factory instance
         */
        @JvmStatic
        fun getInstance(): NamingServiceFactory {
            return INSTANCE
        }
    }
}
